src/InputAreaWidget.ts

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+'use strict';
+
+import * as CodeMirror from 'codemirror';
+
+import 'codemirror/lib/codemirror.css';
+
+import 'codemirror/mode/meta';
+
+import * as dmp from 'diff-match-patch';
+
+import {
+  Message
+} from 'phosphor-messaging';
+
+import {
+  IChangedArgs
+} from 'phosphor-properties';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  ResizeMessage, Widget, Panel
+} from 'phosphor-widget';
+
+import {
+  IInputAreaViewModel, ITextEditorViewModel
+} from './InputAreaViewModel';
+
+
+let diffMatchPatch = new dmp.diff_match_patch()
+
+
+/**
+ * A widget which hosts a CodeMirror editor.
+ */
+export
+class CodeMirrorWidget extends Widget {
+
+  /**
+   * Construct a CodeMirror widget.
+   */
+  constructor(model: ITextEditorViewModel) {
+    super();
+    this.addClass('jp-CodeMirrorWidget');
+    this._model = model;
+    this._editor = CodeMirror(this.node);
+    this.updateMimetype(model.mimetype);
+    this.updateLineNumbers(model.lineNumbers);
+    this.updateFixedHeight(model.fixedHeight);
+    this.updateText(model.text);
+
+    this._editor.on('change', (instance, change) => {
+      this._model.text = this._editor.getDoc().getValue();
+    });
+    model.stateChanged.connect(this._modelUpdate, this);
+  }
+
+  /**
+   * Update whether the editor has a fixed maximum height.
+   */
+  updateFixedHeight(fixedHeight: boolean) {
+    if (fixedHeight) {
+      this.addClass('jp-CodeMirrorWidget-fixedHeight');
+    } else {
+      this.removeClass('jp-CodeMirrorWidget-fixedHeight');
+    }
+  }
+
+  /**
+   * Update the text in the widget.
+   *
+   * #### Notes
+   * This function attempts to restore the cursor to the correct
+   * place by using the bitap algorithm to find the corresponding
+   * position of the cursor in the new text.
+   */
+  updateText(text: string) {
```

Comment:
should all of these `update*` methods be either `protected` or `private`? Is there a strong use case for making them public?